api/core/tools/provider/builtin/databaseoperation/tools/databasecontrol_sqlexec.py

```diff
@@ -0,0 +1,142 @@
+import json
+from typing import Any, Union
+
+import psycopg2
+import pymysql
+from pymysql.constants import CLIENT
+
+from core.tools.entities.tool_entities import ToolInvokeMessage
+from core.tools.tool.builtin_tool import BuiltinTool
+
+
+class DatabaseControlSqlExecTool(BuiltinTool):
+    def _invoke(self,
+                user_id: str,
+                tool_parameters: dict[str, Any]
+                ) -> Union[ToolInvokeMessage, list[ToolInvokeMessage]]:
+        sql = tool_parameters.get('sql')
+        output_format = tool_parameters.get('output_format', 'json')
+        dbsystem = tool_parameters.get('dbsystem')
+        host = tool_parameters.get('host')
+        port = tool_parameters.get('port')
+        username = tool_parameters.get('username')
+        password = tool_parameters.get('password')
+        dbname = tool_parameters.get('dbname')
+
+        result_str = ''
+        if dbsystem == 'postgresql':
+            result_str = self._pg_exec(sql, output_format, host, port, username, password, dbname)
+        elif dbsystem == 'mysql':
+            result_str = self._mysql_exec(sql, output_format, host, port, username, password, dbname)
+        else:
+            raise ValueError(f"Unsupported database system: {dbsystem}")
+
+        return self.create_text_message(result_str)
+
+    def _pg_exec(self, sql: str, output_format: str, host: str, port: str,
+                 username: str, password: str, dbname: str) -> str:
+        result_str = ''
+
+        conn = psycopg2.connect(
+            database=dbname,
+            user=username,
+            password=password,
+            host=host,
+            port=port
+        )
+        cur = conn.cursor()
+        cur.execute(sql)
+        conn.commit()
+
+        if cur.description is not None:
+            fields_tuple = cur.description
+            fields_list = [i[0] for i in fields_tuple]
+
+            if cur.rowcount == 0:
+                result_str = self._convert_to_markdown_table(fields_list, []) \
+                    if output_format == 'markdown' else '[]'
+            else:
+                result_data_list = cur.fetchall()
+                result_list = [dict(zip(fields_list, data_tuple)) for data_tuple in result_data_list]
+                if output_format == 'markdown':
+                    result_str = self._convert_to_markdown_table(fields_list, result_list)
+                else:
+                    result_str = json.dumps(result_list)
+        else:
+            result_str = ''
+
+        cur.close()
+        conn.close()
```

Comment:
Make user the opened resources are properly closed. Use `with` or `try...finally` instead.